wizard/core/state.py

```diff
@@ -0,0 +1,144 @@
+"""
+
+import dataclasses
+from typing import Type, TypeVar
+
+import streamlit as st
+
+import mlcroissant as mlc
+
+T = TypeVar('T', bound='Parent')
+
+
+
+
+class CurrentStep:
+    """hold all major state variables for the application"""
+    start = "start"
+    load = "load"
+    editor = "editor"
+
+
+@dataclasses.dataclass
+class FileObject:
+    """FileObject analogue for editor"""
+    name: str | None = None
+    description: str | None = None
+    content_size: str | None = None
+    content_url: str | None = None
+    encoding_format: str | None = None
+    sha256: str | None = None
+    df: str | None = None
+
+@dataclasses.dataclass
+class FileSet:
+    """FileSet analogue for editor"""
+    contained_in: list[str] = dataclasses.field(default_factory=list)
+    description: str | None = None
+    encoding_format: str | None = ""
+    includes: str | None = ""
+    name: str = ""
+
+@dataclasses.dataclass
+class Field:
+    """Field analogue for editor"""
+    name: str | None = None
+    description: str | None = None
+    data_type: str | None = None
+
+@dataclasses.dataclass
+class RecordSet:
+    """Record Set analogue for editor"""
+    name: str = ""
+    description: str | None = None
+    is_enumeration: bool | None = None
+    key: str | list[str] | None = None
+    fields: list[Field] = dataclasses.field(default_factory=list)
+
+    
+@dataclasses.dataclass
+class Metadata:
+
+    """main croissant data object, helper functions exist to load and unload this into the mlcroissant version"""
+    name: str = ""
+    description: str | None = None
+    citation: str | None = None
+    license: str | None = ""
+    url: str = ""
+    distribution: list[FileObject | FileSet] = dataclasses.field(default_factory=list)
+    record_sets: list[RecordSet] = dataclasses.field(default_factory=list)
+
+    def __bool__(self):
+        return self.name != "" and self.url != ""
+
+    def update_metadata(self, description: str , citation: str , license: license , url: str = "", name: str = "") -> None:
+        self.name = name
+        self.description = description
+        self.citation = citation
+        self.license = license
+        self.url = url
+    def add_distribution(self, distribution: FileSet | FileObject) -> None:
+        self.distribution.append(distribution)
+    def update_distribution(self, key: int, distribution: FileSet | FileObject) -> None:
+        self.distribution[key] = distribution
+    def remove_distribution(self, key: int) -> None:
+        del self.distribution[key]
+    def add_record_set(self, record_set: RecordSet) -> None:
+        self.record_sets.append(record_set)
+    def update_record_set(self, key: int, record_set: RecordSet) -> None:
+        self.record_sets[key] = record_set
+    def remove_record_set(self, key: int) -> None:
+        del self.record_sets[key]
+
+    @classmethod
+    def from_canonical(cls: Type[T], dataset: mlc.Dataset) -> T:
```

Comment:
```
def from_canonical(cls, dataset: mlc.Dataset) -> Metadata:
```

Otherwise, `T` is a generic here.

You may also need:

```python
from __future__ import annotations
```